torch/nn/attention/_flex_attention.py

```diff
@@ -39,11 +39,119 @@ def _identity(
+    BLOCKSPARSE_KV: int
+    BLOCKSPARSE_Q: int
+
+    def __init__(
+        self,
+        kv_num_blocks,
+        kv_indices,
+        q_num_blocks,
+        q_indices,
+        BLOCKSPARSE_KV=_DEFAULT_BLOCKSPARSE_SIZE,
+        BLOCKSPARSE_Q=_DEFAULT_BLOCKSPARSE_SIZE,
+    ):
+        self.kv_num_blocks = kv_num_blocks
+        self.kv_indices = kv_indices
+        self.q_num_blocks = q_num_blocks
+        self.q_indices = q_indices
+        self.BLOCKSPARSE_KV = BLOCKSPARSE_KV
+        self.BLOCKSPARSE_Q = BLOCKSPARSE_Q
+
+
+def broadcast_to_dim(x, dim):
+    while x.dim() < dim:
+        x = x.unsqueeze(0)
+    return x
+
+
+def _convert_mask_to_block_mask(
+    mask,
+    BLOCKSPARSE_KV=_DEFAULT_BLOCKSPARSE_SIZE,
+    BLOCKSPARSE_Q=_DEFAULT_BLOCKSPARSE_SIZE,
+):
+    assert mask.dtype == torch.bool
+    mask = broadcast_to_dim(mask, 4)
+    B, H, Q, KV = mask.shape
+    assert Q % BLOCKSPARSE_Q == 0
+    assert KV % BLOCKSPARSE_KV == 0
+    mask = mask.view(
+        B, H, Q // BLOCKSPARSE_Q, BLOCKSPARSE_Q, KV // BLOCKSPARSE_KV, BLOCKSPARSE_KV
+    )  # [B, H, Q//BLOCKSPARSE_Q, BLOCKSPARSE_Q, KV//BLOCKSPARSE_KV, BLOCKSPARSE_KV]
+    mask = mask.permute(
+        0, 1, 2, 4, 3, 5
+    )  # [B, H, Q//BLOCKSPARSE_Q, KV//BLOCKSPARSE_KV, BLOCKSPARSE_Q, BLOCKSPARSE_KV]
+    mask = mask.sum(dim=[-2, -1]) > 0  # [B, H, Q//BLOCKSPARSE_Q, KV//BLOCKSPARSE_KV]
+    return mask
+
+
+def _convert_block_mask_to_mask(
+    block_mask,
+    BLOCKSPARSE_KV=_DEFAULT_BLOCKSPARSE_SIZE,
+    BLOCKSPARSE_Q=_DEFAULT_BLOCKSPARSE_SIZE,
+):
+    assert block_mask.dim() == 4
+    B, H, Q, KV = block_mask.shape
+    block_mask = block_mask.expand(BLOCKSPARSE_Q, BLOCKSPARSE_KV, *block_mask.shape)
+    block_mask = block_mask.permute(2, 3, 4, 0, 5, 1).reshape(
+        B, H, Q * BLOCKSPARSE_Q, KV * BLOCKSPARSE_KV
+    )
+    return block_mask
+
+
+def _create_block_sparse_mask(
+    mask: torch.Tensor,
+    BLOCKSPARSE_KV: int = _DEFAULT_BLOCKSPARSE_SIZE,
+    BLOCKSPARSE_Q: int = _DEFAULT_BLOCKSPARSE_SIZE,
+):
+    block_mask = _convert_mask_to_block_mask(
+        mask, BLOCKSPARSE_KV=BLOCKSPARSE_KV, BLOCKSPARSE_Q=BLOCKSPARSE_Q
+    )
+    block_mask = block_mask.to(dtype=torch.int8)
+    kv_num_blocks = block_mask.sum(dim=3)
+    kv_indices = torch.argsort(block_mask, dim=3, descending=True, stable=True)
+    q_num_blocks = block_mask.sum(dim=2)
+    q_indices = torch.argsort(block_mask, dim=2, descending=True, stable=True).permute(
+        0, 1, 3, 2
+    )
+    return _BlockSparseMask(
+        kv_num_blocks=kv_num_blocks.to(torch.int32).to(mask.device).contiguous(),
+        kv_indices=kv_indices.to(torch.int32).to(mask.device).contiguous(),
+        q_num_blocks=q_num_blocks.to(torch.int32).to(mask.device).contiguous(),
+        q_indices=q_indices.to(torch.int32).to(mask.device).contiguous(),
+        BLOCKSPARSE_KV=BLOCKSPARSE_KV,
+        BLOCKSPARSE_Q=BLOCKSPARSE_Q,
+    )
+
+
+def _create_empty_block_sparse_mask(query, key, value):
```

Comment:
Add a comment about how this works.

Reply:
I think the comment should be as a doc string, with args, type hints, and a usage section.

Specifically if we are thinking that `_create_block_sparse_mask` is going to be the main user facing API we should explain how we expect it to use. TBH I am not even sure how it should be use and what format I would pass a mask in

Reply:
Oh, I think we don't plan to expose this as public. The public API should be a ```mask_func``` and we infer the block sparse mask from the mask func.